### PR TITLE
Add additional tests for intra expression inference

### LINKED ISSUE #45255: Probable bug: Can't infer an invariant generic on contravariant position when involving another generic
# Bug Report

### 🔎 Search Terms

invariant generic, contravariant inference
Maybe related or perhaps because of the same design limitation mentioned in #44999 

### 🕗 Version & Regression Information

Tested with 4.3.5

### ⏯ Playground Link

[Playground](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEYD2A7AzgF3sAXPAPACoA08AqgHwAUA+ngN7wBueJ8AtnlRqwJTwBeCuVIAzLgFc8ZfkOZIAlsHgBfWcKaLgAKFCRYCZOixQ8AIihnd4aHESpM8AEbmnV3VXrb4zPFGLeHHhYchgBPuLwEoLCXj4+ECBYGCCYAIzmloJRgSraatoecb7w-oGc8NyZZurwYYGR0XLFCUl1qRgATNXZErn5vIXAnoEspeFB8DQxzhONMy3wickdAMyuZr39BUWjfhMVVLVOc3hNsYGtK5gALBtbPnlqQA)

### 💻 Code

```ts
declare const d: <T, U>(_: { v: T, m: (t: T) => U, f: (u: U) => void }) => void
declare const a: "a"
declare const b: "b"

d({
  v: a,
  m: t => t,
  f: u => {
    let test1: "a" = u // Type 'unknown' is not assignable to type '"a"'.
  }
})

d({
  v: a,
  m: (t: "a") => t,
  f: u => {
    let test2: "a" = u
  }
})

d({
  v: a,
  m: _ => b,
  f: u => {
    let test3: "b" = u // Type 'unknown' is not assignable to type '"b"'
  }
})

d({
  v: a,
  m: () => b,
  f: u => {
    let test4: "b" = u
  }
})
```

### 🙁 Actual behavior

In 1st and 3rd function call `U` gets inferred to `unknown`

### 🙂 Expected behavior

In 1st function call  `U` should get inferred to `"a"` and in 3rd function call `U` should get inferred to `"b"`.

Clearly the contravariant position of `U` in `m` seems to be a problem because if we annotate it (as done in 2nd call) or omit it (as done in 4th call) it gets inferred to what is expect and there are no error.

A real world use case:
```ts
declare const branch:
  <T, U extends T>(_: { test: T, if: (t: T) => t is U, then: (u: U) => void }) => void

declare const x: "a" | "b"

branch({
  test: x,
  if: (t): t is "a" => t === "a",
  then: u => {
    let test1: "a" = u // Type '"a" | "b"' is not assignable to type '"a"'
  }
})

branch({
  test: x,
  if: (t: "a" | "b"): t is "a" => t === "a",
  then: u => {
    let test2: "a" = u // compiles
  }
})
```

### PR DESCRIPTION
**Pull request by Andarist**
related to #48538
closes #45255

### CODE REVIEW

File: tests/baselines/reference/intraExpressionInferences.errors.txt [hunk 1]
```diff
@@ -97,6 +97,14 @@ tests/cases/conformance/types/typeRelationships/typeInference/intraExpressionInf
         }
     });
     
+    test({
```
Comment (Andarist) [inline tests/baselines/reference/intraExpressionInferences.errors.txt:100]: #45255 has 2 failing test cases, the 3rd call is basically the same as the call above this one here, but the 1st call (the one that is being added here) is slightly different - it "transfers" the return type from `a` to the `c` parameter type through the identity `b`. So it computes `R1`and `R2` to be the same.

Perhaps it's not that different from the other test case but maybe it's worth adding - i will let you be the judge of that 😉

File: tests/cases/conformance/types/typeRelationships/typeInference/intraExpressionInferences.ts [hunk 2]
```diff
@@ -92,6 +92,14 @@ test({
     }
 });
 
+test({
+    a: () => 0,
+    b: (a) => a,
+    c: (b) => {
+        const x: number = b;
+    }
+});
```
Comment (Andarist) [inline tests/cases/conformance/types/typeRelationships/typeInference/intraExpressionInferences.ts:101]: #45255 has 2 failing test cases, the 3rd call is basically the same as the call above this one here, but the 1st call (the one that is being added here) is slightly different - it "transfers" the return type from `a` to the `c` parameter type through the identity `b`. So it computes `R1`and `R2` to be the same.

Perhaps it's not that different from the other test case but maybe it's worth adding - i will let you be the judge of that 😉

File: tests/cases/conformance/types/typeRelationships/typeInference/intraExpressionInferences.ts [hunk 3]
```diff
@@ -176,3 +184,18 @@ example({
...
+
+declare const x: "a" | "b"
+
+branch({
+  test: x,
+  if: (t): t is "a" => t === "a",
+  then: u => {
+    let test1: "a" = u
+  }
+})
```
Comment (Andarist) [inline tests/cases/conformance/types/typeRelationships/typeInference/intraExpressionInferences.ts:201]: This is very similar to some other cases here but this has an interesting return type annotation at `if`. Previously `u` in `then` was inferred to be of type `'a' | 'b'` but now it gets correctly narrowed down to `'a'`

**Review by DanielRosenwasser — APPROVED** [review]
These seem like interesting test-cases. Thanks!

**Pull request merged**